crates/apollo-compiler/src/validation/fragment.rs

```diff
@@ -10,12 +16,253 @@ pub fn validate_fragment_definitions(
+                                name: name.into(),
+                                ty: type_cond.into(),
+                            },
+                        )
+                        .label(Label::new(
+                            def.loc(),
+                            format!("`{type_cond}` is defined here"),
+                        )),
+                    );
+                }
+            }
 
-        let fragment_type_def = db.find_type_definition_by_name(def.type_condition().to_string());
-        // Make sure the fragment type exists in the schema
-        if fragment_type_def.is_some() {
-            // TODO handle cases where the type does not support fragments (Enum, Scalar...)
-            diagnostics.extend(db.validate_selection_set(def.selection_set().clone()));
+            for inline_def in def.selection_set().inline_fragments().iter() {
+                let inline_type_cond = inline_def.type_condition().unwrap_or_default();
+
+                if BUILT_IN_SCALARS.contains(&inline_type_cond)
+                    && !detected_inline_scalars.contains(inline_type_cond)
+                {
+                    let inline_name = def.name().to_string();
+                    if invalid_inline_fragments.insert((inline_name.clone(), inline_type_cond.to_owned())) {
+                        detected_inline_scalars.insert(inline_type_cond.to_owned());
+                        fragment_diagnostics.push(
+                            ApolloDiagnostic::new(
+                                db,
+                                inline_def.loc().into(),
+                                DiagnosticData::InvalidFragmentType {
+                                    name: inline_name,
+                                    ty: inline_type_cond.into(),
+                                },
+                            )
+                            .label(Label::new(
+                                inline_def.loc(),
+                                format!("`{inline_type_cond}` is defined here"),
+                            )),
+                        );
+                    }
+                }
+            }
+
+            fragment_diagnostics
+        })
+        .collect();
+
+    diagnostics
+}
+
+pub fn validate_fragment_used(
+    db: &dyn ValidationDatabase,
+    file_id: FileId,
+) -> Vec<ApolloDiagnostic> {
+    let mut diagnostics = Vec::new();
+
+    let fragments = db.fragments(file_id);
+
+    let operations = db.operations(file_id);
+    let op_len = operations.len();
+
+    // Fragments must be used within the schema
+    //
+    // Returns Unused Fragment error.
+    if op_len > 0 {
+        for op in operations.iter() {
+            let fields = op.fields(db.upcast());
+            let fragment_spreads = fields
+                .iter()
+                .flat_map(|f| f.selection_set().fragment_spreads())
+                .map(|f| f.name().to_owned())
+                .collect::<Vec<_>>();
```

Comment:
Oooh I absolutely agree -- this would be extremely helpful in reducing the logic when trying to find fragments -- I haven't taken a deep look into how to implement something like this in the HIR but I will do so and ping you for some pairing on this!